components/cdc/src/service.rs

```diff
@@ -68,48 +72,69 @@ impl CdcEvent {
     }
 }
 
+impl fmt::Debug for CdcEvent {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CdcEvent::ResolvedTs(ref r) => {
+                let mut d = f.debug_struct("ResolvedTs");
+                d.field("resolved ts", &r.ts);
+                d.field("region count", &r.regions.len());
+                d.finish()
+            }
+            CdcEvent::Event(e) => {
+                let mut d = f.debug_struct("Event");
+                d.field("region_id", &e.region_id);
+                d.field("request_id", &e.request_id);
+                // prost does not generate (has|get)_entries :(
+                #[cfg(not(feature = "prost-codec"))]
+                if e.has_entries() {
+                    d.field("entries count", &e.get_entries().get_entries().len());
+                }
+                d.finish()
+            }
+        }
+    }
+}
+
 struct EventBatcher {
-    events: Vec<ChangeDataEvent>,
+    buffer: Vec<ChangeDataEvent>,
     last_size: u32,
 }
 
 impl EventBatcher {
-    pub fn with_capacity(cap: usize) -> EventBatcher {
+    fn with_capacity(cap: usize) -> EventBatcher {
         EventBatcher {
-            events: Vec::with_capacity(cap),
+            buffer: Vec::with_capacity(cap),
             last_size: 0,
         }
     }
 
     // The size of the response should not exceed CDC_MAX_RESP_SIZE.
     // Split the events into multiple responses by CDC_MAX_RESP_SIZE here.
-    pub fn push(&mut self, event: CdcEvent) {
+    fn push(&mut self, event: CdcEvent) {
         let size = event.size();
-        if self.events.is_empty() || self.last_size + size >= CDC_MAX_RESP_SIZE {
-            self.last_size = 0;
-            self.events.push(ChangeDataEvent::default());
+        if size >= CDC_MAX_RESP_SIZE {
+            warn!("cdc event too large"; "size" => size, "event" => ?event);
         }
         match event {
             CdcEvent::Event(e) => {
+                if self.buffer.is_empty() || self.last_size + size >= CDC_MAX_RESP_SIZE {
+                    self.last_size = 0;
+                    self.buffer.push(ChangeDataEvent::default());
+                }
                 self.last_size += size;
-                self.events.last_mut().unwrap().mut_events().push(e);
+                self.buffer.last_mut().unwrap().mut_events().push(e);
             }
             CdcEvent::ResolvedTs(r) => {
                 let mut change_data_event = ChangeDataEvent::default();
                 change_data_event.set_resolved_ts(r);
-                self.events.push(change_data_event);
-                // Set last_size to MAX-1 for sending resolved ts as an individual event.
-                // '-1' is to avoid empty event when the next event is still resolved ts.
-                self.last_size = CDC_MAX_RESP_SIZE - 1;
+                self.buffer.push(change_data_event);
```

Comment:
So is it expected that when `ResolvedTs` is followed by `Event`s, it may append `Event`s to the last item in buffer, which already has `ResolvedTs`?